src/main.rs

```diff
@@ -419,53 +421,44 @@ async fn forward_request(
     };
     let is_streaming = http_response.streaming_strategy.is_some();
     let response = if let Some(streaming_strategy) = http_response.streaming_strategy {
-        let (mut sender, body) = body::Body::channel();
-        let agent = agent.as_ref().clone();
-        sender.send_data(Bytes::from(http_response.body)).await?;
-
-        match streaming_strategy {
-            StreamingStrategy::Callback(callback) => {
-                let streaming_canister_id = callback.callback.0.principal;
-                let method_name = callback.callback.0.method;
-                let mut callback_token = callback.token;
-                let logger = logger.clone();
-                tokio::spawn(async move {
-                    let canister = HttpRequestCanister::create(&agent, streaming_canister_id);
-                    // We have not yet called http_request_stream_callback.
-                    let mut count = 0;
-                    loop {
-                        count += 1;
-                        if count > MAX_HTTP_REQUEST_STREAM_CALLBACK_CALL_COUNT {
-                            sender.abort();
-                            break;
-                        }
-
+        let body = http_response.body;
+        let body = futures::stream::once(async move { Ok(body) });
+        let body = match streaming_strategy {
+            StreamingStrategy::Callback(callback) => body::Body::wrap_stream(
+                body.chain(futures::stream::try_unfold(
+                    (
+                        logger.clone(),
+                        agent.clone(),
+                        callback.callback.0,
+                        Some(callback.token),
+                    ),
+                    move |(logger, agent, callback, callback_token)| async move {
+                        let callback_token = if let Some(callback_token) = callback_token {
+                            callback_token
+                        } else {
+                            return Ok(None);
+                        };
```

Comment:
Would something like this be more clear?
```rs
if callback_token.is_none() {
    return Ok(None);
}
let callback_token = callback_token.unwrap();
```

Reply:
`unwrap`ing when you don't have to is very un-Rusty. That said, `match` is more readable than `if let` in this scenario so I will change it.